datafusion/optimizer/src/simplify_expressions/inlist_simplifier.rs

```diff
@@ -0,0 +1,106 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! This module implements a rule that simplifies expressions that is guaranteed to be true or false at planning time
+
+use std::collections::HashSet;
+
+use datafusion_common::tree_node::TreeNodeRewriter;
+use datafusion_common::Result;
+use datafusion_expr::expr::InList;
+use datafusion_expr::{lit, BinaryExpr, Expr, Operator};
+
+/// Simplify expressions that is guaranteed to be true or false to a literal boolean expression
+///
+/// Rules:
+/// If both expressions are positive or negative, then we can apply intersection of both inlist expressions
+///     1. `a in (1,2,3) AND a in (4,5) -> a in (1,2,3,4,5)`
+///     2. `a not int (1,2,3) AND a not in (4,5,6) -> a not in (1,2,3,4,5,6)`
+/// If one of the expressions is negated, then we apply exception on positive expression
+///     1. `a in (1,2,3) AND a not in (1,2,3,4,5) -> a in ()`
```

Comment:
Shouldn't this be rewritten to `false`?
```suggestion
///     1. `a in (1,2,3) AND a not in (1,2,3,4,5) -> false`
```

Reply:
Yes, I thought `a in ()` is more understandable in this context